### PR TITLE
Resolves #578 (partially): lazy-play landing page videos

### LINKED ISSUE #578: Need to Examine Memory and CPU Usage With New Landing Page and Optimize Accordingly
We need to examine the memory and CPU usage of the new landing page and optimize assets accordingly.

### PR DESCRIPTION
**Pull request by andrew4699**
Resolves #578 
(Partially)

- Landing page videos now pause when out of view
- New instruction video cycling code (those videos no longer cycle when out of view)

To test;
- Ensure that landing page videos still appear to be always playing (users shouldn't see them paused)
- Ensure that instruction videos (1/2/3) still cycle
- Observe at least some performance gain

### CODE REVIEW

**Comment by lstrobel** [comment]
Will review when I have time - midterms/snow are really messing with my workload

**Comment by misaugstad** [comment]
>Observe at least some performance gain

@andrew4699 how to you suggest to observe performance gain? Of course it is hard to measure here, but are you looking for the page to load slightly faster? Or for there to be less lag on the page when scrolling?

**Comment by paarig** [comment]
@andrew4699 Tested and it appears to be working as designed. Any idea how I can check for a performance improvement?

**Comment by andrew4699** [comment]
When you scroll out of view of the hero video, scrolling through the rest of the page should be faster relative to the current live website. This also applies to the 3 instruction videos below the hero video.

**Comment by misaugstad** [comment]
>scrolling through the rest of the page should be faster relative to the current live website

For those testing, you want to be comparing it to your local machine running the develop branch, not necessarily the live website.

**Comment by misaugstad** [comment]
This definitely seems to help. Using lazy play, here is what my CPU usage looked like:

cpu-lazyplay

The first section CPU usage is ~50%, and that is at the top of the page with the "Let's create a path for everyone" video playing. The second section has CPU usage ~40%, and that is when I've scrolled to the area with the laptop instructional videos playing. The last part has CPU usage ~20%, and here I've scrolled below that to where no video is actively playing.

To compare with develop, here is my CPU usage as I scrolled down the entire page on the develop branch. CPU usage stays ~50% the entire time.

cpu-develop

Nice work @andrew4699 :sunglasses:

File: public/assets/homepage.js [hunk 1]
```diff
@@ -192,4 +197,71 @@ $( document ).ready(function() {
...
+    // Setup video lazyPlay
+    $(window).on("scroll", onScroll);
+
+    vidBanner = $('#vidbanner')[0];
+    bannerVid = $('#bgvid')[0];
+
+    instructVideoContainer = $('#instructionvideo')[0];
+    instructVideos = [$('#vid1')[0],
+        $('#vid2')[0],
+        $('#vid3')[0]];
```
Comment (misaugstad) [inline public/assets/homepage.js]: I would go with either
```
instructVideos = [$('#vid1')[0], $('#vid2')[0], $('#vid3')[0]];
```
or
```
instructVideos = [
    $('#vid1')[0],
    $('#vid2')[0],
    $('#vid3')[0]
];
```

**Comment by jonfroehlich** [comment]
Woohooo!

File: public/assets/homepage.js [hunk 2]
```diff
@@ -192,4 +197,71 @@ $( document ).ready(function() {
...
+function onScroll() {
+    lazyPlayVideosThrottled(); // While scrolling, run the check every 300ms
+    lazyPlayVideosDebounced(); // After scrolling, make sure we run the check
+}
+
+// lazyPlays our main videos
+function lazyPlayVideos() {
+    lazyPlay(vidBanner, bannerVid);
+    return;
+    for(var i = 0; i < instructVideos.length; i++) {
```
Comment (misaugstad) [inline public/assets/homepage.js]: I would keep a space between the `for` and the `(`. In general, I would suggest having a space before an open paren _unless_ it is a function call. So `for (...)` and `if (...)`, but `lazyPlay(...)` and `myFunc(...)`.

File: public/assets/homepage.js [hunk 3]
```diff
@@ -192,4 +197,71 @@ $( document ).ready(function() {
...
+        lazyPlay(instructVideoContainer, instructVideos[i]);
+    }
+}
+
+// Pauses a video if a certain element is outside of the viewport.
+// Plays the video otherwise.
+function lazyPlay(el, video) {
+    console.log("inView = " + isElementVerticallyVisible(el), "isPlaying = " + isVideoPlaying(video));
```
Comment (misaugstad) [inline public/assets/homepage.js]: Just a reminder to remove the console statement before I merge.

File: public/assets/homepage.js [hunk 4]
```diff
@@ -192,4 +197,71 @@ $( document ).ready(function() {
...
+
+    if(isElementVerticallyVisible(el)) {
```
Comment (misaugstad) [inline public/assets/homepage.js]: Same spacing comment with these `if`s.

File: public/assets/homepage.js [hunk 5]
```diff
@@ -192,4 +197,71 @@ $( document ).ready(function() {
...
+    return !pausedVideos[video];
+}
+
+// Returns true if the given element is in the vertical viewport
+function isElementVerticallyVisible(el) {
+    var rect = el.getBoundingClientRect();
+    var windowHeight = (window.innerHeight || document.documentElement.clientHeight);
+
+    return (rect.top <= windowHeight) && ((rect.top + rect.height) >= 0);
+}
```
Comment (misaugstad) [inline public/assets/homepage.js]: Aaand could you just add a newline to the end of this file?

File: public/assets/homepage.js (hunk 2 above)
Comment (andrew4699) [reply public/assets/homepage.js]: I also forgot to remove the "return;" there.

**Comment by andrew4699** [comment]
Just fixed the style issues.

**Comment by misaugstad** [comment]
Looks great, thanks!

**Pull request merged**